### DISCUSSION TITLE
Recommended way of accessing Clipboard in Mvvm?

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#11170] [Q&A]
Clipboard has moved from Application to TopLevel in Preview7.

So it breaks the following registration in my code. 

    services.Register<IClipboard>(() => Application.Current!.Clipboard);

What's the recommended way of accessing IClipboard in a Mvvm design? Should I look for ApplicationLifetime and get the MainWindow or MainView depending on mobile or desktop, or is there a more recommended way to do it?

**Comment by jp2masa** [comment]
If you follow MVVM strictly, you will want to create a platform-independent abstraction of the clipboard, provided by all viewmodels that represent a `TopLevel`. This would then require you to pass the clipboard down to child viewmodels as required (you can also have an abstraction for all `TopLevel` services so that you only have to edit viewmodel classes once, similar to passing an `IServicesProvider`). Alternatively, if you use DI with scopes, you can inject the clipboard in the scope associated with the `TopLevel`.

Another way (hacky) is to create an attached property that simply returns the clipboard (eventually with a converter from Avalonia to your abstraction). This would still work better than using `ApplicationLifetime` in my opinion.

**Reply by mysteryx93** [reply]
Note that `IClipboard` is already a platform-independent abstraction. I don't see the point of re-wrapping it unless I ever plan on using something else than Avalonia. At which point it would be easy to replace.

btw why was the clipboard changed to `TopLevel` to begin with? Is there a difference between using clipboard reference from one window or from another window?

**Reply by jp2masa** [reply]
In that case, you're not following MVVM strictly, but if you plan to use only Avalonia it should be fine.

> btw why was the clipboard changed to TopLevel to begin with? Is there a difference between using clipboard reference from one window or from another window?

Probably because of Android:

https://github.com/AvaloniaUI/Avalonia/blob/9f64bad365880fa12d29d17eaecd68b92504f35a/src/Android/Avalonia.Android/Platform/SkiaPlatform/TopLevelImpl.cs#L57

**Reply by maxkatz6** [reply]
Also, we want to avoid static/global API where possible.

**Reply by mysteryx93** [reply]
ok re-reading your answer.

> If you follow MVVM strictly, you will want to create a platform-independent abstraction of the clipboard

Optional, but good for strict MVVM.

> provided by all viewmodels that represent a TopLevel.

> This would then require you to pass the clipboard down to child viewmodels as required

Passing it similar to passing the "owner" ViewModel to child view models to display dialogs with MvvmDialogs.

**The question is how does that Clipboard service gets set in the main ViewModel?**

So far, everything is written with 0 code-behind.

> (you can also have an abstraction for all TopLevel services so that you only have to edit viewmodel classes once, similar to passing an IServicesProvider).

Optional but good idea if there are more needs; but kind of competes with the constructor DI principles.

> Alternatively, if you use DI with scopes, you can inject the clipboard in the scope associated with the TopLevel.

There is no scopes with Splat.

> Another way (hacky) is to create an attached property that simply returns the clipboard (eventually with a converter from Avalonia to your abstraction). This would still work better than using ApplicationLifetime in my opinion.

Attached property and write-only binding? I don't like that Dependency Injection matters rely on XAML code on each view.

**Reply by jp2masa** [reply]
You can't avoid code-behind. Even converters can be considered code-behind. You should simply keep it to a minimum.

The best option for you is probably having an `Initialize` method in the viewmodel, which the view calls on data context changed, to pass the required services (ReactiveUI implements similar patterns).

**Reply by mysteryx93** [reply]
I've come all this way with 0 code-behind, managing dialogs and all that. I don't like this solution. There's got to be a cleaner/simpler way.

**Reply by mysteryx93** [reply]
It's a bit frustrating because it was working perfectly fine in preview6. If it was changed for Android, it only has a single VisualRoot to begin with, so I'm really not sure of any benefit. From a code perspective, I just want to use `IClipboard` like any other input/output injected services. Unless there's a clear pitfall to it.

**Reply by workgroupengineering** [reply]
`IStorageProvider` also has a similar problem. @wieslawsoltes suggested [this](https://github.com/AvaloniaUI/Avalonia/discussions/11093#discussioncomment-5686977) way.

@mysteryx93 I agree 1000% with you. I think that interfaces like `IClipboard` and `IStorageProvider` must be accessible at the `Application` level because they are not strictly linked to the view like `IInsetsManager` which in multi monitor scenario could assume different values according to the monitor in which the `TopLevel` is present.

**Reply by mysteryx93** [reply]
@wieslawsoltes is taking the route of looking into `ApplicationLifetime`

        if (Application.Current?.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime { MainWindow: { } window })
        {
            return window.Clipboard;
        }

        if (Application.Current?.ApplicationLifetime is ISingleViewApplicationLifetime { MainView: { } mainView })
        {
            var visualRoot = mainView.GetVisualRoot();
            if (visualRoot is TopLevel topLevel)
            {
                return topLevel.Clipboard;
            }
        }

The way I solved it with `IStorageProvider` is that MvvmDialogs handles it. Unless you want to use an API that MvvmDialogs is not exposing.

**Reply by mysteryx93** [reply]
Unless there's a clear downside to it, I think I'll opt for an extension method that returns the TopLevel from the Application class.

```c#
public static TopLevel? GetTopLevel(this Application app)
{
    if (app.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
    {
        return desktop.MainWindow;
    }
    if (app.ApplicationLifetime is ISingleViewApplicationLifetime viewApp)
    {
        var visualRoot = viewApp.MainView?.GetVisualRoot();
        return visualRoot as TopLevel;
    }
    return null;
}
```

Remains the problem that the TopLevel is not yet initialized when creating the first ViewModel, thus the service can't be injected properly.

Unless I create a wrapper service around it that requests the underlying service when needed.

**Reply by mysteryx93** [reply]
And my `ClipboardService`. Unless I run into any problems, I'll use that for now.

```
public class ClipboardService : IClipboardService
{
    protected virtual IClipboard? Clipboard => _clipboard ??= Application.Current.GetTopLevel()?.Clipboard;
    private IClipboard? _clipboard;

    public Task<string?> GetTextAsync() => Clipboard?.GetTextAsync() ?? Task.FromResult<string?>(null);
    public Task SetTextAsync(string? text) => Clipboard?.SetTextAsync(text) ?? Task.CompletedTask;
    public Task ClearAsync() => Clipboard?.ClearAsync() ?? Task.CompletedTask;
    public Task SetDataObjectAsync(IDataObject data) => Clipboard?.SetDataObjectAsync(data) ?? Task.CompletedTask;
    public Task<string[]> GetFormatsAsync() => Clipboard?.GetFormatsAsync() ?? Task.FromResult(Array.Empty<string>());
    public Task<object?> GetDataAsync(string format) => Clipboard?.GetDataAsync(format) ?? Task.FromResult<object?>(null);
}
```

btw I see that Avalonia is lacking in documentation not only for end-user but also in the code. Clipboard classes have no documentation whatsoever.

**Reply by kierepka** [reply]
Not working in 11 
Avalonia.Application.Current.GetTopLevel()?.Clipboard 
Error CS1061 : „Application” missing „GetTopLevel”

**Reply by mysteryx93** [reply]
Add this 

```
public static class ApplicationExtensions
{
    /// <summary>
    /// Returns the TopLevel from the main window or view. 
    /// </summary>
    /// <param name="app">The application to get the TopLevel for.</param>
    /// <returns>A TopLevel object.</returns>
    public static TopLevel? GetTopLevel(this Application? app)
    {
        if (app?.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
        {
            return desktop.MainWindow;
        }
        if (app?.ApplicationLifetime is ISingleViewApplicationLifetime viewApp)
        {
            var visualRoot = viewApp.MainView?.GetVisualRoot();
            return visualRoot as TopLevel;
        }
        return null;
    }
}
```

**Reply by zkxjzmswkwl** [reply]
This is comical.

**Comment by dioptryk** [comment]
I believe all of the above ugly static TopLevel/Application references are unnecessary. There's an elegant solution to the problem of injecting platform-dependent services to ViewModel, before the Window is initialized. Observe (I'm using Microsoft.Extensions.DependencyInjection here):

```cs
public interface IPlatformServicesAccessor
{
    IStorageProvider? StorageProvider { get; }
}

class DefaultPlatformServiceAccessor : IPlatformServicesAccessor
{
    readonly IClassicDesktopStyleApplicationLifetime _desktop;

    public DefaultPlatformServiceAccessor(IClassicDesktopStyleApplicationLifetime desktop)
    {
        _desktop = desktop;
    }

    public IStorageProvider? StorageProvider => _desktop.MainWindow?.StorageProvider;
}
```

Then, register the services and construct the Window (App.xaml.cs):

```cs
var services = new ServiceCollection();

services.AddSingleton<IPlatformServicesAccessor>(new DefaultPlatformServiceAccessor(desktop));
services.AddTransient<MyViewModel>();

var provider = services.BuildServiceProvider();

desktop.MainWindow = new MyMainWindow
{
    DataContext = provider.GetRequiredService<MyViewModel>()
};
```

And then, in MyViewModel you simply inject:
```cs
public MyViewModel(IPlatformServicesAccessor platformServices)
{
    _platformServices = platformServices;
}
```

And use:
```cs
[RelayCommand]
private async void SaveMyFile()
{
    var result = await _platformServices.StorageProvider.SaveFilePickerAsync(new());
    // your logic
}
```

Of course, you don't have to expose IStorageProvider (or IClipboard or whatever).. you can abstract to your heart's content.
